From ecf46aba63ada3616f6a192287f3f1680f48d8d4 Mon Sep 17 00:00:00 2001
From: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>
Date: Thu, 14 Jul 2011 22:51:21 +0200
Bug-Debian: https://bugs.debian.org/633998
FIXME: This patch needs to be updated to latest upstream version
Subject: [PATCH 1/3] fix constant mac addresses in case they are specified
 more than once

I have
|pre-up macchanger -m 00:00:00:00:00:01 wlan0

in my network script. The first ifup works, the second fails with:

|mikejones:~# ifup wlan0
|Permanent MAC: xx:xx:xx:xx:xx:xx (xxx)
|Current   MAC: 00:00:00:00:00:01 (unknown)
|It's the same MAC!!
|Failed to bring up wlan0.

Signed-off-by: Sebastian Andrzej Siewior <sebastian@breakpoint.cc>
---
 src/main.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/main.c b/src/main.c
index 9a58937..64c96e2 100644
--- a/src/main.c
+++ b/src/main.c
@@ -93,6 +93,7 @@ main (int argc, char *argv[])
 	char permanent    = 0;
 	char print_list   = 0;
 	char show         = 0;
+	char same_okay    = 0;
 	char *set_mac     = NULL;
 	char *search_word = NULL;
 	
@@ -208,6 +209,7 @@ main (int argc, char *argv[])
 		exit (0);
 	} else if (set_mac) {
 		if (mc_mac_read_string (mac_faked, set_mac) < 0) exit(1);
+		same_okay = 1;
 	} else if (random) {
 		mc_mac_random (mac_faked, 6);
 	} else if (endding) {
@@ -228,10 +230,13 @@ main (int argc, char *argv[])
 	}
 
 	/* Is the same MAC? */
-	if (mc_mac_equal (mac, mac_faked)) {
-		printf ("It's the same MAC!!\n");
+	if (!same_okay && mc_mac_equal (mac, mac_faked)) {
+		if (!same_okay) {
+			printf ("It's the same MAC!!\n");
 
-		ret = 1;
+			ret = 1;
+		} else
+			ret = 0;
 	} else {
 		/* Set the new MAC */
 		if (mc_net_info_set_mac (net, mac_faked) >= 0) {
-- 
1.7.5.4

